### PR TITLE
MOBILE-1009 Fix handle data for GridField app context

### PR DESCRIPTION
**Pull request by mclomello**
Inconsistence between data changed before debounce and prepared values for request submit fixed

### CODE REVIEW

File: android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/FormInstanceRecord.kt [hunk 1]
```diff
@@ -21,6 +21,8 @@ import com.liferay.apio.consumer.model.Thing
 import com.liferay.apio.consumer.model.get
 import com.liferay.mobile.screens.ddl.form.util.FormConstants
 import com.liferay.mobile.screens.ddl.model.Field
+import org.json.JSONArray
+import org.json.JSONObject
```
Comment (phcp) [inline android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/FormInstanceRecord.kt]: Remove unused imports.

File: android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/GridField.kt [hunk 2]
```diff
@@ -28,15 +29,21 @@ class GridField : Field<Grid>, Parcelable {
...
+    }
+
+    override fun convertToFormattedString(value: Grid?): String {
+        return value?.rawValues.let {
+			gson.toJson(it)
+        }.toString()
+    }
 
-	constructor(parcel: Parcel, classLoader: ClassLoader) : super(parcel, classLoader) {
+    constructor(parcel: Parcel, classLoader: ClassLoader) : super(parcel, classLoader) {
```
Comment (phcp) [inline android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/GridField.kt]: SF all the changes from this class.

File: android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormPresenter.kt [hunk 3]
```diff
@@ -29,7 +29,7 @@ import java.util.*
  */
 class DDMFormPresenter(val view: DDMFormViewContract.DDMFormView) : DDMFormViewContract.DDMFormViewPresenter {
 
-	private val interactor = DDMFormInteractor()
+    private val interactor = DDMFormInteractor()
```
Comment (phcp) [inline android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormPresenter.kt]: SF

File: android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormPresenter.kt [hunk 4]
```diff
@@ -76,28 +76,33 @@ class DDMFormPresenter(val view: DDMFormViewContract.DDMFormView) : DDMFormViewC
...
 				it.fieldValues[field.name] = field.toData() ?: ""
 			}
+		}
+	}
 
-			if (!view.hasConnectivity()) {
-				view.showOfflineWarningMessage()
-				return
-			}
+    override fun onDebounceSync(thing: Thing, formInstance: FormInstance, field: Field<*>) {
```
Comment (phcp) [inline android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormPresenter.kt]: Rename onDebounceSync to a meaningful name. Maybe we should rename all sync methods as the suggestions above (or something similar):
- onFieldValueChanged: syncFieldModels
- onDebounceSync: syncForm
- syncFormInstance: syncInitialContext

But these changes should be in another PR.

File: android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormView.kt [hunk 5]
```diff
@@ -247,13 +247,17 @@ class DDMFormView @JvmOverloads constructor(
...
+            presenter.onFieldValueChanged(it)
+        }.debounce(2, TimeUnit.SECONDS)
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe({ field ->
+                thing?.let {
+                    presenter.onDebounceSync(it, formInstance, field)
+                }
+            }, {
+                LiferayLogger.e(it.message)
+            })
```
Comment (phcp) [inline android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormView.kt]: SF

File: android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormView.kt [hunk 6]
```diff
@@ -247,13 +247,17 @@ class DDMFormView @JvmOverloads constructor(
...
-			}
-		}
-
-		subscription.add(autoSaveSubscription)
+        subscription = observable.doOnNext {
+            presenter.onFieldValueChanged(it)
+        }.debounce(2, TimeUnit.SECONDS)
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe({ field ->
+                thing?.let {
```
Comment (phcp) [inline android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormView.kt]: An exception should be thrown if thing is null. You could do something similar as the following code:
```
val thing = thing ?: throw Exception("No thing found")
```

**Review by phcp — CHANGES_REQUESTED** [review]
It's a good approach, thanks @mclomello for the PR! 
Please review the changes that I suggested, just some small things.

File: android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormPresenter.kt (hunk 4 above)
Comment (victorlaerte) [reply android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormPresenter.kt]: We can follow presenter guidelines, for actions, we can use `action + subject` e.g.: `loadForm`, `loadInitialContext`

And for events the `subject + actionTriggered` e.g: `fieldValueChanged`, `formRecordChanged` or `formContextChanged`

We can discuss this better if you guys want.

**Review by victorlaerte — CHANGES_REQUESTED** [review]

File: android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/FormInstanceRecord.kt (hunk 1 above)
Comment (mclomello) [reply android/library/src/main/java/com/liferay/mobile/screens/ddm/form/model/FormInstanceRecord.kt]: Imports removed like mentioned

File: android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormPresenter.kt (hunk 4 above)
Comment (mclomello) [reply android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormPresenter.kt]: Guys, like you both mentioned, I changed the method names following the Presenter Guideline that suggested by Victor.
Now the methods names are more readable and clear.

File: android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormView.kt (hunk 6 above)
Comment (mclomello) [reply android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormView.kt]: Thanks Paulo. I added the treatment like was mentioned.

File: android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormView.kt [hunk 7]
```diff
@@ -246,15 +246,19 @@ class DDMFormView @JvmOverloads constructor(
...
+        subscription = observable.doOnNext {
+            presenter.fieldModelsChanged(it)
+        }.debounce(2, TimeUnit.SECONDS)
+        .observeOn(AndroidSchedulers.mainThread())
+        .subscribe({ field ->
+            thing?.let {
+                presenter.syncForm(it, formInstance, field)
+            } ?: throw Exception("No thing found")
+        }, {
+            LiferayLogger.e(it.message)
```
Comment (victorlaerte) [inline android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/ddm/form/DDMFormView.kt]: To a better Log message, you should use `LiferayLogger.e(it.message, e)`

**Review by victorlaerte — APPROVED** [review]

**Comment by phcp** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*
